src/crypto/openssl/cose_verifier.cpp

```diff
@@ -223,4 +228,113 @@ namespace ccf::crypto
   {
     return std::make_unique<COSEKeyVerifier_OpenSSL>(public_key);
   }
+
+  COSEEndorsementValidity extract_cose_endorsement_validity(
+    std::span<const uint8_t> cose_msg)
+  {
+    UsefulBufC msg{cose_msg.data(), cose_msg.size()};
+    QCBORError qcbor_result;
+
+    QCBORDecodeContext ctx;
+    QCBORDecode_Init(&ctx, msg, QCBOR_DECODE_MODE_NORMAL);
+
+    QCBORDecode_EnterArray(&ctx, nullptr);
+    qcbor_result = QCBORDecode_GetError(&ctx);
+    if (qcbor_result != QCBOR_SUCCESS)
+    {
+      LOG_DEBUG_FMT("Failed to parse COSE_Sign1 outer array");
+      return {};
+    }
+
+    const uint64_t tag = QCBORDecode_GetNthTagOfLast(&ctx, 0);
+    if (tag != CBOR_TAG_COSE_SIGN1)
+    {
+      LOG_DEBUG_FMT("Failed to parse COSE_Sign1 tag");
+      return {};
```

Comment:
I suggest switching to exceptions for this, and deciding whether to log at the level above. The log messages are going to be very difficult to correlate with a particular transaction/context this way. Returning a default value also creates the risk of misuse at the call site.